### PR TITLE
Leaderboard frontend

### LINKED ISSUE #9: Add Leaderboard - High Score
As a player, I want to see how well I play compared to other players by seeing a list of player scores after I complete my game so that I can gloat about it.

### PR DESCRIPTION
**Pull request by mtwete**
Finished leaderboard frontend logic, pulled main into the branch so all functionality is present. NOTE: the fake leaderboard score file is still full, I will empty it once the input name branch is done

### CODE REVIEW

File: constants.py [hunk 1]
```diff
@@ -22,9 +28,23 @@
 HEX_BACKGROUND = os.path.join(IMAGES_DIR, 'Hex-Background-Resized.jpg')
 BACKGROUND_IMAGES_FILE_PATHS = [MENU_BACKGROUND, HEX_BACKGROUND]
 
+#Leaderboard static variables
+LEADERBOARD_SCORE_FILE_PATH = os.path.join(MAIN_DIR, "resources/leaderboard_scores.txt")
+LEADERBOARD_TITLE = 'TOP SCORES:'
```
Comment (minavu) [inline constants.py:33]: These are your only changes to this file yet it's marking other lines as well.  I wonder why.

File: leaderboard.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    def draw(self):
+        self.image.blit(self.image_original, self.rect)
+        self.button_group.update()
+        self.button_group.draw(self.image)
+
+    #Method to check for button clicks
+    def check_button_click(self):
+        #Check for exit button or main menu click, if there's nothing stay at the leaderboard
+        return self.return_button.check_click() or self.exit_button.check_click() or LEADER_BOARD
+
```
Comment (minavu) [inline leaderboard.py:46]: Nice work with this.  It looks a lot like the menu class.  If we have more time, we can refactor an abstract class or something.

File: leaderboard_line_writer.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        # so replace the tabs with 4 spaces
+        reformatted_line = line.replace('\t', ' ' * 4)
+        score_line = self.font.render(reformatted_line, True, 'white')
+        # Set up proper position, so they are space right
+        position = score_line.get_rect()
+        position.centerx = self.background_rect_width / 2
+        position.y = (self.background_rect_height / 10)+(self.leaderboard_title_rect_height / 2)+self.line*self.line_height
+        # Add the score line to the list of the score lines
+        self.leaderboard_score_lines.append((score_line, position))
+        self.line += 1
```
Comment (minavu) [inline leaderboard_line_writer.py:61]: Nice job with this too.  Again, if we have more time, we can refactor.

File: long_text_writer.py [hunk 4]
```diff
@@ -2,18 +2,17 @@
 
 class LongTextWriter(pygame.sprite.Sprite):
 
-    def __init__(self, long_text, font='courier', size=16, color='white', width=300):
+    def __init__(self, long_text, font='courier', font_size=16, color='white', width=300):
```
Comment (minavu) [inline long_text_writer.py]: Good rename!  Thanks!

File: missile.py [hunk 5]
```diff
@@ -1,23 +1,48 @@
...
         self.x -= int(self.x_vel)
         self.y -= int(self.y_vel)
-        pygame.draw.circle(display, (255, 255, 255), (self.x, self.y), 2)
\ No newline at end of file
+        pygame.draw.circle(display, (255, 255, 255), (self.x, self.y), 2)
+
+    #update bullet image while in motion
+    def update(self):
+        self.rect.move_ip(-int(self.x_vel), -int(self.y_vel))
+        pygame.draw.circle(self.image, self.color, self.image.get_rect().center, 2)
```
Comment (minavu) [inline missile.py]: This file was already reviewed in a previous PR....

File: player.py [hunk 6]
```diff
@@ -3,41 +3,63 @@
...
+            self.rect.x = 0
+        if self.rect.x >= 766:
+            self.rect.x = 766
+        if self.rect.y <= 0:
+            self.rect.y = 0
+        if self.rect.y >= 566:
+            self.rect.y = 566 
 
     def update_score(self, add_point=1):
         self.score += add_point
```
Comment (minavu) [inline player.py]: Same with this one.

File: tests/test_leaderboard_line_writer.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        self.leaderboard_line_writer.render_score_line.assert_has_calls(calls, any_order=False)
+
+    def test_render_score_line(self):
+        lines = self.leaderboard_line_writer.leaderboard_scores_text.splitlines()
+        old_line_num = self.leaderboard_line_writer.line
+        old_leaderboard_score_lines_len = len(self.leaderboard_line_writer.leaderboard_score_lines)
+        self.leaderboard_line_writer.render_score_line(lines[0])
+        self.assertEqual(self.leaderboard_line_writer.line,old_line_num+1)
+        self.assertEqual(len(self.leaderboard_line_writer.leaderboard_score_lines), old_leaderboard_score_lines_len+1)
+
```
Comment (minavu) [inline tests/test_leaderboard_line_writer.py:73]: Wow, great work thinking up so many test cases!

**Review by minavu — APPROVED** [review]
A lot of the changes in this PR are from previous PRs that are already merged, so I didn't comment on them.  I would just say to be careful when you merge because something looks wonky here but IDKY it's like this.

File: missile.py (hunk 5 above)
Comment (mtwete) [reply missile.py]: Sorry I think there was an issue when I tried to merge the branch

File: constants.py (hunk 1 above)
Comment (mtwete) [reply constants.py:33]: Yeah I don't know either, I think something got screwed up on my end

File: leaderboard.py (hunk 2 above)
Comment (mtwete) [reply leaderboard.py:46]: Yeah I was thinking the same, just didn't have the time to do it

File: leaderboard_line_writer.py (hunk 3 above)
Comment (mtwete) [reply leaderboard_line_writer.py:61]: Yep, at least it is a good learning experience even if we don't have a chance!

File: long_text_writer.py (hunk 4 above)
Comment (mtwete) [reply long_text_writer.py]: No problem!

File: tests/test_leaderboard_line_writer.py (hunk 7 above)
Comment (mtwete) [reply tests/test_leaderboard_line_writer.py:73]: Thanks! I tried to do as much as I could but it was hard to test!

**Comment by mtwete** [comment]
Yeah I'm not sure why the file thing happened, maybe it had something to do with that issue I had with my IDE?
Anyways thanks for the review!

**Pull request merged**